torch/csrc/jit/passes/quantization.cpp

```diff
@@ -1,14 +1,84 @@
+#include <stack>
 #include <torch/csrc/jit/passes/quantization.h>
 
 #include <torch/csrc/jit/ir.h>
 #include <torch/csrc/jit/node_hashing.h>
 #include <torch/csrc/jit/passes/alias_analysis.h>
 
-#include <stack>
+#include <unordered_set>
 
 namespace torch {
 namespace jit {
+namespace {
+// QuantizerUtils
+
+int checkIfNodeValuesQuantizable(Node* n) {
+  // This is map for quantizable nodes. It will be expanded in future to
+  // support more ops and patterns.
+  static const std::unordered_set<std::string> nodeLookup =
+   {"aten::_convolution", "aten::relu", "aten::conv2d"};
+  AT_ASSERT(n != nullptr);
+  return nodeLookup.count(n->kind().toQualString());
+}
+
+bool isTypeTensor(Value *v) {
+  // Quantizer will look at all Tensor subTypes of inputs and outputs
+  // for quantization
+  return(v->isCompleteTensor() || v->type()->kind() == TypeKind::TensorType
+    || v->type()->kind() == TypeKind::DimensionedTensorType);
+}
+
+// Check if Value external input to block
+bool checkifExternalInput(Value* v) {
+  return v->node()->kind().toQualString() == std::string("prim::Param");
+}
+
+// Add Quant-Dequant node pair for quantizable node
+void addQuantDeQuantNodes(Value* v, Node* n=nullptr) {
+  // This flag tells whether Value is external input or not. This
+  // determine insertion point for quant-dequant node.
+  bool external = true;
+  if (n == nullptr) {
+    // If caller supplies node, then Value is external param because v->node()
+    // does not belong to any block and insertion point is before supplied
+    // node. Else Value is activation and insertion happens after v->node()
+    n = v->node();
+    external = false;
+  }
+  // Create quant node
+  Node* quant =
+      n->owningGraph()->create(at::Symbol::fromQualString("aten::quant"));
+  AT_ASSERT(quant != nullptr);
+  quant->outputs()[0]->setUniqueName(v->uniqueName() + ".quant");
+
+  // Create dequant node
+  Node* dequant =
+      n->owningGraph()->create(at::Symbol::fromQualString("aten::dequant"));
+  AT_ASSERT(dequant != nullptr);
+  dequant->outputs()[0]->setUniqueName(v->uniqueName() + ".dequant");
+
+  // Add quant-dequant nodes to the graph at appropriate insertion points
+  if (!external) {
```

Comment:
This can use a comment describing what the before/after graph should look like for both the parameter and non-parameter case. It is hard for me to know if this is correct because I do not know the intent.

I would also suggested splitting this function into a Param and non-Param function, because the way it edits the graph is different for each case.

Reply:
I agree that splitting the function would help.